src/lib/core/CHIPTLVReader.cpp

```diff
@@ -128,83 +128,112 @@ CHIP_ERROR TLVReader::Get(bool & v)
 CHIP_ERROR TLVReader::Get(int8_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(int8_t));
     v              = CastToSigned(static_cast<uint8_t>(v64));
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(int16_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(int16_t));
     v              = CastToSigned(static_cast<uint16_t>(v64));
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(int32_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(int32_t));
     v              = CastToSigned(static_cast<uint32_t>(v64));
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(int64_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(int64_t));
     v              = CastToSigned(v64);
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(uint8_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(uint8_t));
     v              = static_cast<uint8_t>(v64);
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(uint16_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(uint16_t));
     v              = static_cast<uint16_t>(v64);
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(uint32_t & v)
 {
     uint64_t v64   = 0;
-    CHIP_ERROR err = Get(v64);
+    CHIP_ERROR err = Get(v64, sizeof(uint32_t));
     v              = static_cast<uint32_t>(v64);
     return err;
 }
 
 CHIP_ERROR TLVReader::Get(uint64_t & v)
 {
+    uint64_t v64   = 0;
+    CHIP_ERROR err = Get(v64, sizeof(uint64_t));
+    v              = v64;
+    return err;
+}
+
+static inline CHIP_ERROR VerifySize(size_t requestedSize, size_t actualSize)
+{
+    return (requestedSize < actualSize) ? CHIP_ERROR_WRONG_TLV_TYPE : CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TLVReader::Get(uint64_t & v, size_t size) const
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
     switch (ElementType())
     {
     case TLVElementType::Int8:
         v = static_cast<uint64_t>(static_cast<int64_t>(CastToSigned(static_cast<uint8_t>(mElemLenOrVal))));
         break;
     case TLVElementType::Int16:
-        v = static_cast<uint64_t>(static_cast<int64_t>(CastToSigned(static_cast<uint16_t>(mElemLenOrVal))));
+        if ((error = VerifySize(size, sizeof(int16_t))) == CHIP_NO_ERROR)
```

Comment:
Will do.

#### Minority report :)
IMHO, the whole thing is a bit over engineered: I'd rather say -- the only legitimate decoding is to the type of **T**LV, any further  conversation is purely caller's business.

Reply:
Per the TLV spec the "type of TLV" is not really well-defined.  For the same exact schema and the same exact value being sent the type can be any integer type that is big enough to fit the value.  So if you try to be more restrictive than that when reading, you will fail to parse valid TLV payloads.

Reply:
@bzbarsky-apple: Aren't we control both encoding and decoding ends? If not, we may have a serious issue here: I see no byte order (endianness) handling. Am I missing something?

Reply:
> Aren't we control both encoding and decoding ends?

No, we do not, generally speaking.  The ends both have to follow the spec, but they don't have to be using this specific implementation.

> I see no byte order (endianness) handling. Am I missing something?

Spec says integers are encoded little-endian in TLV.  In terms of implementation, on the reader side, the endianness handling is in `TLVReader::ReadElement` where `mElemLenOrVal` is set to things like `LittleEndian::Read16(p)` (and so forth, depending on the size).  So `mElemLenOrVal` is in native endianness, and directly manipulating it in `Get` is fine.

On the writer side, `TLVWriter::WriteElementHead` does things like `LittleEndian::Write16(p, static_cast<uint16_t>(lenOrVal));` so we're ok there too, afaict.

Reply:
> In terms of implementation, on the reader side, the endianness handling is in TLVReader::ReadElement where mElemLenOrVal is set to things like LittleEndian::Read16(p) (and so forth, depending on the size). So mElemLenOrVal is in native endianness, and directly manipulating it in Get is fine.

Oops, missed int, indeed. Thanks

Reply:
> 1. If the caller asks for a signed type and the TLV type is unsigned, or vice versa, error out.
> 2. If the actual value that was encoded does not fit in the caller's requested type (this sort of thing can be checked via `CanCastTo`) error out.

How is 1. not a subset of 2.?  If the TLV number can fit into the caller-requested type, nothing breaks.  Why not allow this?

Reply:
Hence my "I'm not sure what we should do about signedness checks" above.

I think the key is basically that the TLV schema _does_ pin down whether a value should be sent as signed vs unsigned, so using the wrong signedness on the sender is an error.  But it does not pin down encoding size, so using a larger-than-strictly-needed encoding is allowed encoder behavior.

Those are all spec-side decisions that could be revisited, of course, but given the state of the spec right now we end up with code behavior where we check that signedness matches and then check that the value fits in the requested range....

Reply:
I honestly don't see how this is a spec-side consideration.  When I see things like this, I begin to wonder whether we all have computer brain and have forgotten that numbers are natural things that can be abstracted from their representation.

Here we are converting from TLV types to C types.  Sometimes a number represented in the tlv cannot fit inside the C type, so we have to range-check before performing the conversion.  This works just as well for negative numbers as for any other potential overflow.  And it is negative numbers specifically, not signed numbers in general, that cannot fit inside an unsigned C type.  Why then are we special-casing signedness?

FWIW, I'm basically just reframing this rationale:

> Say the encoder encoded the value "5" as a `TLVElementType::Int16`. This is a perfectly valid thing to do.
> 
> The decoder expects a `uint8_t`. This check will fail, since `size` will be 1, but it should succeed, because the value is in fact perfectly valid as a `uint8_t`.

What was wrong with that?  It's simple and internally consistent.

Reply:
You're making the argument that the spec should allow either SIGNED INTEGER or UNSIGNED INTEGER encoding for a nonnegative integer, up to the sender.  That's fine as an argument, but then it should be made in a spec issue.  Right now the spec does not allow that (and considers a TLV invalid if there's a signed/unsigned mismatch), and does allow choosing the size you encode as.  To the extent that we're implementing the spec here we should implement that behavior.

Reply:
I'm not though.  I gather that the spec requires tlv type to match for any exchanges.  I'm not arguing against that.  I'm arguing that this is not a good place to enforce this because it should be the prerogative of an implementer to choose internal representations (and particularly if those representations can hold the desired range of values).

And throwing these errors here enshrines the following:

If my code is trying to convert to a uint8:
  -> this implies that there is an external interface of unsigned TLV type.  Does it though?
If my code tries to convert from a signed TLV to a uint8:
  -> this means somebody violated an external interface contract.  Are we sure this is true?

It seems odd to me to enforce external interface contracts based upon the type of a variable the code happens to pass to these getters.  What if we foolishly publish an interface that requires a signed type and restricts range to [0,255]?  Thrifty implementer sees this and uses a uint8_t  internally in firmware.  Now this code will emit a runtime error.  Where's the bug?  In the spec?  Or in the firmware?  I'm arguing that there should be no bug at all because nothing actually needs to break for this case.

I'll admit though that this approach is likely to achieve the desired type-enforcement.  I'm completely OK not winning this argument.  I only come at this with some similar experiences around CBOR, where I had to argue pretty vigorously with other embedded C folks that my preferred 'if it fits, it ships' approach was the right one.  I know it feels wrong coming from a world of type safety.  But it can be really great.  I'll defer to the other reviewers on this one though.

Reply:
It seems my previous comment disappeared somehow, so my apology if this is a duplicate.

I have updated the PR, implementing the rules suggested by @bzbarsky-apple.
@bzbarsky-apple, I'd appreciate your feedback, approval or disapproval :)

Towards the discussion:

On the sender side, I'd say, we need to enforce whatever the specs imply. Outside of that, I tend to agree with 'if it fits, it ships' maxima, spelling fits as in "fits sender's needs'.

On the receiver side, I still think the only legitimate decoding should be to the TLV's internal type. IMHO, all valuable-guessing is beyond TLV scope. Basically, it's just 'we receive what was shipped' approach. I don't see how it would break anything as we just shift the valuable-guessing responsibility towards the receiver.

Reply:
> I'm arguing that this is not a good place to enforce this because it should be the prerogative of an implementer to choose internal representations

That is fair.   Maybe we can put the signed/unsigned enforcement in some higher-level schema-validation step, then assuming the schema passes just use whatever types as long as the data fits in them.